core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -868,8 +872,8 @@ protected static RexNode substitute(RexCall call, int ordinal, RexNode node) {
     }
 
     private RexNode visit(final RexNode call) {
-      int i = reducibleExps.indexOf(call);
-      if (i == -1) {
+      Integer i = reducibleExpsMap.get(call);
+      if (i == null) {
```

Comment:
Why not just use a `HashSet` ?

Reply:
the value `i` indicates the index to find value in `reducedValues`

Reply:
Oops, my fault.